### PR TITLE
Implement filesystem operations interface for containers in test Harness

### PR DESCRIPTION
**Pull request by weiiwang01**
This pull request addresses the enhancement request [#931](https://github.com/canonical/operator/issues/931) for an API that allows access to the virtual filesystem in the test module. 

The current testing module of the operator framework includes a mock filesystem for the pebble-controlled container but lack an easy way to manipulate and inspect the mock filesystem in tests directly. To solve this, we've added a `get_container_filesystem` method to the test `Harness` class. This method provides an interface for conducting filesystem operations on a specified container, without restrictions posed by `set_can_connect`.

Changelog:

 - Adds `Harness.get_container_filesystem` method, allowing for direct filesystem operations on a specified test container without dealing with `set_can_connect`.

### CODE REVIEW

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2669]: Any reason to extract these to locals rather than just use eg `self._container` directly?

File: test/test_testing.py
Comment (benhoyt) [inline test/test_testing.py:4336]: Thanks for adding tests! We should probably also add one that checks that the protocol's method signatures exactly match the `model.Container` methods. Should be able to loop through the protocol methods and use `inspect.signature` and `==` to check? (I haven't tested this.)

**Review by benhoyt — CHANGES_REQUESTED** [review]
Thanks for this! I've left a few minor comments and one structural one for discussion.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: I rename the `_TestingContainerFilesystemFacade` class to `ContainerFilesystem`.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Updated, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Updated, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Updated, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Updated to the new docstring format.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Removed, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Updated, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Updated, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: No reason for that, I just updated the function to use `self._container` directly.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2703]: I don't think cache will be very helpful here, since the `_TestingContainerFilesystemProxy` shares a similar lifecycle as `Harness`, so cache will be destroyed at the end of each test case, lowering the hit rate significantly.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Yes, the problem with manually passing all the arguments is that the majority of methods use many arguments. Manually passing them is prone to error, and such errors can be very difficult to find.

File: ops/testing.py
Comment (benhoyt) [reply ops/testing.py:2718]: Sounds reasonable, thanks

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:242]: Instead of a `type: ignore` comment, how about using `cast()`? Also, I'm not sure passing the args as keyword args adds much here.

```suggestion
        proxy = _TestingContainerFilesystemProxy(container, self._backend)
        return cast(ContainerFilesystem, proxy)
```

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2615]: Per [the Protocol PEP 544](https://peps.python.org/pep-0544/), it's best if abstract methods `raise NotImplementedError` in their body. It probably doesn't matter much here as we're always using the proxy, but just in case we accidentally instantiate/subclass this, might as well do that on all these methods.

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2698]: Naming nit: Let's drop the `Testing` here, and just make this `_ContainerFilesystemProxy` to more closely relate to `ContainerFilesystem`.

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2715]: Let's do a `try ... finally` here just in case the test is doing some exception handling, to avoid `can_connect` remaining in an unexpected state. I think that would look like this:

```suggestion
            try:
                result = func(*args, **kwargs)
            finally:
                self._backend._set_can_connect(self._container._pebble, can_connect)
```

File: test/test_testing.py
Comment (benhoyt) [inline test/test_testing.py:4377]: Nice and simple, thanks for adding!

File: test/test_testing.py
Comment (benhoyt) [inline test/test_testing.py:4378]: When you iterate a `dict` directly, you get keys by default, so `.keys()` is not usually necessary.

However, here I'd use `for name, func in ContainerFilesystem.__dict__.items()` and then use `func` below instead of `getattr(ContainerFilesystem, attr)`.

File: ops/testing.py
Comment (benhoyt) [inline ops/testing.py:2620]: Sorry if I put you on the wrong track earlier, but these need the full `ops.model...` prefix to be linked correctly in the generated docs. Try `tox -e docs` and then `open docs/_build/html/index.html` to test it.

**Review by benhoyt — CHANGES_REQUESTED** [review]
I quite like how this turned out -- thank you! Just a few minor comments/changes requested. After that I'll approve and get a second approval from John M or similar.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Updated, I check the document and it's linked correctly now.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Oops, that's a big one, thank you for spotting this, fixed! I also added a test case for the exception safety.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Renamed, thanks!

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2614]: Updated, although I have to complete some type hints in the `ops.model.Container` for type checks to pass.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py]: Updated to use `cast` instead, thanks!

**Review by benhoyt — APPROVED** [review]
Great, thanks for those fixes -- approving!

File: ops/testing.py
Comment (PietroPasotti) [inline ops/testing.py:2614]: IIRC, with Protocols, you don't have to explicitly implement the methods, nor make them abstract, since you're not supposed to subclass a Protocol.

Typically you do
```
class Foo(Protocol):
    def bar(self, a: str, b:int, **) -> MyReturnType: ...
```

No docstrings, no raise NotImplementedError... just an Ellipsis will do.
Again, nothing will subclass this, so the docstrings are just 'wasted'. What a protocol defines is the 'api' of an object.

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
I see a few typing patterns that I'd do differently. Other than that, the API looks fine.

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:2614]: (if you want to preserve docstrings on your proxy object, you should 'copy them' over from `model.Container` 
like:
```python
def _copy_docstrings(source_cls: Any):
    """Copy the docstrings from source_cls to target_cls.

    Use this as:
      @_copy_docstrings(source_class)
      class TargetClass:

    And for any public method that exists on both classes, it will copy the
    __doc__ for that method.
    """
    def decorator(target_cls: Any):
        for meth_name, _ in target_cls.__dict__.items():
            if meth_name.startswith('_'):
                continue
            source_method = source_cls.__dict__.get(meth_name)
            if source_method is not None and source_method.__doc__:
                target_cls.__dict__[meth_name].__doc__ = source_method.__doc__
        return target_cls
    return decorator
```

(from `ops.testing`)

File: ops/model.py
Comment (weiiwang01) [reply ops/model.py:2015]: Pyright requires the `ContainerFilesystem` Protocol to have completed typing (i.e. with return types). So to make checking the consistency between `Container` and `ContainerFilesystem` I added return types to the `Container` class as well.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:241]: I think `return cast(ContainerFilesystem, proxy)` and `return cast(model.Container, proxy)` doesn't have any difference. Since from an outsider's perspective, the return type is `ContainerFilesystem`. The cast here is just make the type checker happy.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2614]: @benhoyt mentioned that in the [the Protocol PEP 544](https://peps.python.org/pep-0544/) examples, the `NotImplementedError` is used in the abstract methods.
Also, here's a collapsed conversation about using `_copy_docstring` https://github.com/canonical/operator/pull/933#discussion_r1199858265

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2715]: The return type of the `get_container_filesystem` is `ops.testing.ContainerFilesystem` instead of  `ops.model.Container`. The `ops.testing.ContainerFilesystem` class just doesn't have other attributes from the definition.

File: ops/testing.py
Comment (PietroPasotti) [reply ops/testing.py:2614]: yeah but do they need to be abstractmethods in the first place?

Comment (weiiwang01) [reply ops/testing.py:2614]: I think in our use case, with or without `abstractmethod` are interchangeable.

File: ops/testing.py
Comment (benhoyt) [reply ops/testing.py:241]: @PietroPasotti Having the return type as `model.Container` and avoiding the protocol would work, but I think it sends the wrong signal to the user -- we only want them to use this to interact with the file system (hence the name `get_container_filesystem`). The idea is to show them it's special access for use only in tests, not a real container which has can-connect behaviour.

File: ops/testing.py
Comment (benhoyt) [reply ops/testing.py:2715]: Yeah, I think what @weiiwang01 is saying is reasonable. However, I don't think it'd hurt to clarify the message. Perhaps something simple like:

```
raise AttributeError(f'object has no attribute {name!r}; get_container_filesystem only provides access to file-related methods')
```

File: ops/model.py
Comment (benhoyt) [reply ops/model.py:2015]: Not sure why this change was required, but an explicit `-> None` does seem to be what's recommended by [PEP 484](https://peps.python.org/pep-0484/). Part of the reason is that type checkers don't check functions which don't have any type annotations, and if you leave `-> None` off, for a function with no arguments like `def foo(): ...` the type checkers wouldn't check it, but they would check `def foo() -> None: ...` because it has type annotations.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2715]: How about changing the error message to `f'{self.__class__.__name__!r} object has no attribute {name!r}'`? This way it will print `'ContainerFilesystem' object has no attribute 'foo'`.

This error message aligns with Python's default message for attribute not found. Moreover, I don't feel like mentioning `get_container_filesystem` in the `ContainerFilesystem` since `ContainerFilesystem` is a dependency of `get_container_filesystem`. So it shouldn't have information about the `get_container_filesystem`.

Comment (benhoyt) [reply ops/testing.py]: Works for me, thanks!

**Comment by benhoyt** [comment]
@weiiwang01 If you could also update this branch by merging in the latest main changes, that would be great.

**Comment by benhoyt** [comment]
> why are we modeling ContainerFilesystem as the subset of Container that lets you push/pull from pebble, rather than having ContainerFilesystem just be "_TestingFilesystem".
>
> Its a much smaller api, it doesn't need magic toggling of Container model state, and it still lets you poke at what is on the "filesystem" of the container, without trying to pretend it is a Container that isn't a Container.

Thanks @jameinel for your detailed review and strong feedback. When I first looked at making `_TestingFilesystem` public it didn't seem like the best API to expose. But maybe it just needs a few tweaks (and of course documentation).

> I'm very much in favor of having an object which can poke the filesystem, but doing so by mutating the model to say that it can is going to bite us in the future. (it doesn't follow the principle of expected behavior.)

The way I see this is that it's an implementation detail: the user doesn't need to know how we're achieving it (and won't know that the model is being changed and then changed back). I agree this seems a bit hacky, but what's the actual concern here, or how would the user know that anything's changing?

In any case, I'll think further about the "Expose TestingFilesystem" approach and play with it a bit next week.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:2725]: Well, I can certainly achieve this without mutating the model state, by temporary hijacking the `_check_connection` in the test pebble client and disabling all the checks for the filesystem access.

```python
        @functools.wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            pebble_client = self._container._pebble
            with unittest.mock.patch.object(target=pebble_client, attribute="_check_connection"):
                return func(*args, **kwargs)
```

Do you think this could make it better?

Comment (weiiwang01) [reply ops/testing.py:2725]: The incentive to mimic the API of the container for the API of the container filesystem is that users will not need to learn a new API just for testing; they can reuse their knowledge of the container API.

File: ops/testing.py
Comment (weiiwang01) [reply ops/testing.py:61]: Yes, since the ops module doesn't have a dedicated typing module for custom types, it has to be imported from where it's defined.

**Comment by benhoyt** [comment]
I can think of three possible approaches here:

1) Have get_container_filesystem(c) return `_TestingFilesystem` (and make that public, possibly trimmed down). Simple to change. However, it's a large new (public) API surface, and I'm not sure the `_TestingFilesystem` API was thought through well enough to be public-quality.
2) Add something like `Harness.set_filesystem_content(container: str, content: dict[str, StrOrBytes])`, where the content dict is key: file content or similar. Simple, small API surface. More of a declarative API. Similar to how [ops-scenario](https://github.com/canonical/ops-scenario) does it.
3) Add something like `Harness.set_filesystem_source(local_path: str)` where you simply point the testing file system to a local directory.

@weiiwang01 I'm going to schedule some time for us to chat and brainstorm about the best approach.

**Comment by weiiwang01** [comment]
> I can think of three possible approaches here:
> 
> 1. Have get_container_filesystem(c) return `_TestingFilesystem` (and make that public, possibly trimmed down). Simple to change. However, it's a large new (public) API surface, and I'm not sure the `_TestingFilesystem` API was thought through well enough to be public-quality.
> 2. Add something like `Harness.set_filesystem_content(container: str, content: dict[str, StrOrBytes])`, where the content dict is key: file content or similar. Simple, small API surface. More of a declarative API. Similar to how [ops-scenario](https://github.com/canonical/ops-scenario) does it.
> 3. Add something like `Harness.set_filesystem_source(local_path: str)` where you simply point the testing file system to a local directory.
> 
> @weiiwang01 I'm going to schedule some time for us to chat and brainstorm about the best approach.

The challenge with using a dictionary lies in its limited expressiveness in representing a filesystem; it lacks a clear way to differentiate between directories and files, among other issues. Also, there will be needs for inspecting the testing virtual filesystem, which does not work with the dictionary approach.

The difficulty in using the local filesystem resides in the requirement to import it into the testing virtual filesystem. Also, it doesn't work with inspecting the testing virtual filesystem either. 

I would love the opportunity to discuss this new design further in a meeting.

**Comment by benhoyt** [comment]
Update from our video discussion today: @weiiwang01 and I both like the idea of replacing the in-memory filesystem with a real temp directory (which would be managed and cleaned up automatically by Harness). Weii's going to put up a small spec for further discussion, before we go further with the code.

**Comment by benhoyt** [comment]
Per discussion and per @weiiwang01's OP036 spec, closing this in favour of a different approach which uses the real filesytem.

**Pull request closed without merging**
